massa-network/src/peer_info_database.rs

```diff
@@ -823,71 +755,365 @@ impl PeerInfoDatabase {
+            self.get_peer_type(ip).unwrap()
+        };
 
-        let res = match self.peers.entry(*ip) {
-            hash_map::Entry::Occupied(mut occ) => {
-                let peer = occ.get_mut();
-                if (peer.bootstrap
-                    && self.active_bootstrap_connections
-                        >= self.network_settings.target_bootstrap_connections)
-                    || (!peer.bootstrap
-                        && self.active_in_nonbootstrap_connections
-                            >= self.network_settings.max_in_nonbootstrap_connections)
-                {
-                    Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
-                } else if peer.banned {
-                    massa_trace!("in_connection_refused_peer_banned", {"ip": peer.ip});
-                    peer.last_failure = Some(MassaTime::compensated_now(self.clock_compensation)?);
-                    self.request_dump()?;
-                    Err(NetworkConnectionErrorType::BannedPeerTryingToConnect(*ip))
-                } else if peer.active_in_connections
-                    >= self.network_settings.max_in_connections_per_ip
-                {
-                    self.request_dump()?;
-                    Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
-                } else {
-                    peer.active_in_connections += 1;
-                    if peer.bootstrap {
-                        self.active_bootstrap_connections += 1;
-                    } else {
-                        self.active_in_nonbootstrap_connections += 1;
-                    }
-                    Ok(())
-                }
+        if self.is_max_in_connection_count_reached(peer) {
+            return Err(NetworkError::PeerConnectionError(
+                NetworkConnectionErrorType::MaxPeersConnectionReached(*ip),
+            ));
+        }
+
+        let peer = self.peers.get_mut(ip).ok_or({
+            NetworkError::PeerConnectionError(NetworkConnectionErrorType::PeerInfoNotFoundError(
+                *ip,
+            ))
+        })?; // peer was inserted just before
+
+        // is there a attempt slot avaible
+        if peer.banned {
+            massa_trace!("in_connection_refused_peer_banned", {"ip": peer.ip});
+            peer.last_failure = Some(MassaTime::compensated_now(self.clock_compensation)?);
+            self.request_dump()?;
+            return Err(NetworkError::PeerConnectionError(
+                NetworkConnectionErrorType::BannedPeerTryingToConnect(*ip),
+            ));
+        } else if peer.active_in_connections >= self.network_settings.max_in_connections_per_ip {
+            self.request_dump()?;
+            return Err(NetworkError::PeerConnectionError(
+                NetworkConnectionErrorType::MaxPeersConnectionReached(*ip),
+            ));
+        } else {
+            peer.active_in_connections += 1;
+        }
+
+        let peer = self.get_peer_type(ip).ok_or({
```

Comment:
I would do this:
```rust
let peer_type = peer.type;
std::mem::drop(peer);

// here we have peer_type, but not the peer reference in scope anymore
```

Reply:
Yes, you can do that, you also can use scopes.